File: C/Baekjoon/15000/15480.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

int parent[100001][18] = {}, depth[100001] = {};

int LCA(int u, int v)
{
    if (depth[u] < depth[v])
    {
        int temp = v;
        v = u;
        u = temp;
    }

    int diff = depth[u] - depth[v], i = 0;
    while (diff)
    {
        if (diff % 2)
            u = parent[u][i];
        i++;
        diff /= 2;
    }

    for (int i = 17; i >= 0; i--)
    {
        if (parent[u][i] != parent[v][i])
        {
            u = parent[u][i];
            v = parent[v][i];
        }
    }
    if (u != v)
        u = parent[u][0];
    return u;
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    int N, M, r, u, v;
    cin >> N;

    vector<int> arr[N + 1];
    bool checked[N + 1] = {};

    for (int i = 0; i < N - 1; i++)
    {
        cin >> u >> v;
        arr[v].push_back(u);
        arr[u].push_back(v);
    }

    stack<int> st;
    st.push(1);
    checked[1] = true;

    while (!st.empty())
    {
        u = st.top();
        st.pop();

        for (int i : arr[u])
        {
            if (checked[i])
                continue;

            parent[i][0] = u;
            checked[i] = true;
            depth[i] = depth[u] + 1;

            for (int j = 1; 1 << j <= N; j++)
            {
                if (!parent[i][j - 1])
                    break;
                parent[i][j] = parent[parent[i][j - 1]][j - 1];
            }

            st.push(i);
        }
    }

    cin >> M;
    while (M--)
    {
        cin >> r >> u >> v;
        int i = LCA(r, u), j = LCA(u, v), k = LCA(v, r), res;

        if (depth[i] >= depth[j] && depth[i] >= depth[k])
        {
            res = i;
        }
        else if (depth[j] >= depth[i] && depth[j] >= depth[k])
        {
            res = j;
        }
        else
        {
            res = k;
        }

        cout << res << '\n';
    }
}

File: Python/Baekjoon/15000/15480.py
from sys import stdin
from collections import deque

N = int(stdin.readline())
arr = [[] for i in range(N + 1)]

for i in range(N - 1):
    A, B = map(int, stdin.readline().rstrip().split())
    arr[A].append(B)
    arr[B].append(A)

st = deque([1])
checked = [False for i in range(N + 1)]
checked[1] = True

depth = [0 for i in range(N + 1)]
parent = [[0 for j in range(18)] for i in range(N + 1)]

while len(st):
    p = st.pop()
    for q in arr[p]:
        if checked[q]:
            continue

        checked[q] = True
        parent[q][0] = p
        depth[q] = depth[p] + 1

        k = 1
        while 2**k <= N:
            if not parent[q][k - 1]:
                break
            parent[q][k] = parent[parent[q][k - 1]][k - 1]
            k += 1

        st.append(q)


def LCA(A, B):
    if depth[B] > depth[A]:
        A, B = B, A

    diff = depth[A] - depth[B]
    for j in range(17, -1, -1):
        if 2**j <= diff:
            diff -= 2**j
            A = parent[A][j]

    for j in range(17, -1, -1):
        if parent[A][j] != parent[B][j]:
            A, B = parent[A][j], parent[B][j]
    if A != B:
        A = parent[A][0]

    return A


M = int(stdin.readline())
for i in range(M):
    R, U, V = map(int, stdin.readline().rstrip().split())
    A, B, C = LCA(R, U), LCA(R, V), LCA(U, V)

    if depth[A] >= depth[B] and depth[A] >= depth[C]:
        print(A)
    elif depth[B] >= depth[A] and depth[B] >= depth[C]:
        print(B)
    else:
        print(C)
